File: libs/typescript/tests/output.test.ts
import { describe, expect, test, beforeEach, afterEach } from "bun:test";
import {
  writeFileSync,
  mkdirSync,
  existsSync,
  readFileSync,
  readdirSync,
} from "node:fs";
import { join } from "node:path";
import yaml from "js-yaml";
import {
  inferOutputName,
  readBlockManifest,
  writeOutputs,
} from "../src/output.ts";
import {
  File,
  Directory,
  FileCollection,
  RasterFile,
  VectorFile,
  JsonFile,
  RasterFileCollection,
} from "../src/types.ts";
import { setupWorkDir } from "./helpers.ts";

let workDir: { path: string; originalCwd: string };

beforeEach(() => {
  workDir = setupWorkDir();
});

afterEach(() => {
  process.chdir(workDir.originalCwd);
});

describe("inferOutputName", () => {
  test("File", () => {
    expect(inferOutputName(new File("/tmp/a"))).toBe("file");
  });

  test("RasterFile", () => {
    expect(inferOutputName(new RasterFile("/tmp/a"))).toBe("raster");
  });

  test("VectorFile", () => {
    expect(inferOutputName(new VectorFile("/tmp/a"))).toBe("vector");
  });

  test("JsonFile", () => {
    expect(inferOutputName(new JsonFile("/tmp/a"))).toBe("json");
  });

  test("FileCollection", () => {
    expect(inferOutputName(new FileCollection([]))).toBe("files");
  });

  test("RasterFileCollection", () => {
    expect(inferOutputName(new RasterFileCollection([]))).toBe("rasters");
  });
});

describe("writeOutputs", () => {
  test("null result produces no output", () => {
    writeOutputs(null);
    expect(readdirSync(join(workDir.path, "outputs"))).toHaveLength(0);
  });

  test("single RasterFile output", () => {
    const src = join(workDir.path, "tmp_result.tif");
    writeFileSync(src, "raster data");
    writeOutputs(new RasterFile(src));
    const outputDir = join(workDir.path, "outputs", "raster");
    expect(existsSync(outputDir)).toBe(true);
    expect(existsSync(join(outputDir, "tmp_result.tif"))).toBe(true);
    expect(readFileSync(join(outputDir, "tmp_result.tif"), "utf-8")).toBe(
      "raster data"
    );
  });

  test("single file with manifest uses manifest name", () => {
    const src = join(workDir.path, "result.tif");
    writeFileSync(src, "data");
    const manifest = { custom_output: { type: "file", format: "GeoTIFF" } };
    writeOutputs(new RasterFile(src), manifest);
    expect(
      existsSync(join(workDir.path, "outputs", "custom_output", "result.tif"))
    ).toBe(true);
  });

  test("dict output", () => {
    const srcRaster = join(workDir.path, "result.tif");
    writeFileSync(srcRaster, "raster");
    const srcJson = join(workDir.path, "summary.json");
    writeFileSync(srcJson, '{"key": "value"}');

    writeOutputs({
      raster: new RasterFile(srcRaster),
      summary: new JsonFile(srcJson),
    });

    expect(
      existsSync(join(workDir.path, "outputs", "raster", "result.tif"))
    ).toBe(true);
    expect(
      existsSync(join(workDir.path, "outputs", "summary", "summary.json"))
    ).toBe(true);
  });

  test("collection output", () => {
    const paths: string[] = [];
    for (let i = 0; i < 3; i++) {
      const p = join(workDir.path, `tile_${i}.tif`);
      writeFileSync(p, `tile ${i}`);
      paths.push(p);
    }
    writeOutputs(new RasterFileCollection(paths));
    const outputDir = join(workDir.path, "outputs", "rasters");
    expect(existsSync(outputDir)).toBe(true);
    expect(readdirSync(outputDir)).toHaveLength(3);
  });

  test("directory output", () => {
    const srcDir = join(workDir.path, "result_dir");
    mkdirSync(srcDir);
    writeFileSync(join(srcDir, "file1.txt"), "a");
    writeFileSync(join(srcDir, "file2.txt"), "b");

    writeOutputs(new Directory(srcDir));
    const outputDir = join(workDir.path, "outputs", "directory");
    expect(existsSync(outputDir)).toBe(true);
    expect(existsSync(join(outputDir, "file1.txt"))).toBe(true);
    expect(existsSync(join(outputDir, "file2.txt"))).toBe(true);
  });

  test("preserves filename", () => {
    const src = join(workDir.path, "my_custom_name.geojson");
    writeFileSync(src, "geojson data");
    writeOutputs(new VectorFile(src));
    expect(
      existsSync(
        join(workDir.path, "outputs", "vector", "my_custom_name.geojson")
      )
    ).toBe(true);
  });
});

describe("readBlockManifest", () => {
  test("no manifest returns null", () => {
    expect(readBlockManifest()).toBeNull();
  });

  test("block.yaml in CWD", () => {
    const manifest = {
      id: "test.block",
      outputs: { raster: { type: "file", format: "GeoTIFF" } },
    };
    writeFileSync("block.yaml", yaml.dump(manifest));
    const result = readBlockManifest();
    expect(result).toEqual({ raster: { type: "file", format: "GeoTIFF" } });
  });

  test("env var manifest takes priority", () => {
    const manifestPath = join(workDir.path, "external_block.yaml");
    const manifest = { outputs: { output: { type: "file" } } };
    writeFileSync(manifestPath, yaml.dump(manifest));
    const original = process.env["SPADE_BLOCK_MANIFEST"];
    process.env["SPADE_BLOCK_MANIFEST"] = manifestPath;
    try {
      const result = readBlockManifest();
      expect(result).toEqual({ output: { type: "file" } });
    } finally {
      if (original === undefined) {
        delete process.env["SPADE_BLOCK_MANIFEST"];
      } else {
        process.env["SPADE_BLOCK_MANIFEST"] = original;
      }
    }
  });
});

File: libs/go/output_test.go
package spade

import (
	"os"
	"path/filepath"
	"testing"
)

// ---------------------------------------------------------------------------
// WriteOutputs tests
// ---------------------------------------------------------------------------

func TestWriteSingleRasterFile(t *testing.T) {
	dir := setupWorkDir(t)
	src := filepath.Join(dir, "tmp_result.tif")
	os.WriteFile(src, []byte("raster data"), 0644)

	result := NewRasterFile(src)
	if err := WriteOutputs(&result, dir, nil); err != nil {
		t.Fatal(err)
	}

	data, err := os.ReadFile(filepath.Join(dir, "outputs", "raster", "tmp_result.tif"))
	if err != nil {
		t.Fatal(err)
	}
	if string(data) != "raster data" {
		t.Fatalf("expected 'raster data', got %q", string(data))
	}
}

func TestWriteSingleFile(t *testing.T) {
	dir := setupWorkDir(t)
	src := filepath.Join(dir, "result.dat")
	os.WriteFile(src, []byte("file data"), 0644)

	result := NewFile(src)
	if err := WriteOutputs(&result, dir, nil); err != nil {
		t.Fatal(err)
	}

	if _, err := os.Stat(filepath.Join(dir, "outputs", "file", "result.dat")); err != nil {
		t.Fatal("expected output file to exist")
	}
}

func TestWriteRasterFileCollection(t *testing.T) {
	dir := setupWorkDir(t)
	var paths []string
	for i, name := range []string{"tile_0.tif", "tile_1.tif", "tile_2.tif"} {
		p := filepath.Join(dir, name)
		os.WriteFile(p, []byte{byte(i)}, 0644)
		paths = append(paths, p)
	}

	result := NewRasterFileCollection(paths)
	if err := WriteOutputs(&result, dir, nil); err != nil {
		t.Fatal(err)
	}

	entries, _ := os.ReadDir(filepath.Join(dir, "outputs", "rasters"))
	if len(entries) != 3 {
		t.Fatalf("expected 3 files, got %d", len(entries))
	}
}

func TestWriteDirectoryOutput(t *testing.T) {
	dir := setupWorkDir(t)
	srcDir := filepath.Join(dir, "result_dir")
	os.MkdirAll(srcDir, 0755)
	os.WriteFile(filepath.Join(srcDir, "file1.txt"), []byte("a"), 0644)
	os.WriteFile(filepath.Join(srcDir, "file2.txt"), []byte("b"), 0644)

	result := NewDirectory(srcDir)
	if err := WriteOutputs(&result, dir, nil); err != nil {
		t.Fatal(err)
	}

	outputDir := filepath.Join(dir, "outputs", "directory")
	if _, err := os.Stat(filepath.Join(outputDir, "file1.txt")); err != nil {
		t.Fatal("file1.txt not found")
	}
	if _, err := os.Stat(filepath.Join(outputDir, "file2.txt")); err != nil {
		t.Fatal("file2.txt not found")
	}
}

func TestWriteWithManifestCustomName(t *testing.T) {
	dir := setupWorkDir(t)
	src := filepath.Join(dir, "result.tif")
	os.WriteFile(src, []byte("data"), 0644)

	manifest := map[string]any{
		"custom_output": map[string]any{"type": "file"},
	}

	result := NewRasterFile(src)
	if err := WriteOutputs(&result, dir, manifest); err != nil {
		t.Fatal(err)
	}

	if _, err := os.Stat(filepath.Join(dir, "outputs", "custom_output", "result.tif")); err != nil {
		t.Fatal("expected file at custom_output/result.tif")
	}
}

func TestWriteMultipleNamedOutputs(t *testing.T) {
	dir := setupWorkDir(t)
	rasterSrc := filepath.Join(dir, "result.tif")
	os.WriteFile(rasterSrc, []byte("raster"), 0644)
	jsonSrc := filepath.Join(dir, "summary.json")
	os.WriteFile(jsonSrc, []byte(`{"key": "value"}`), 0644)

	outputs := NewOutputs()
	r := NewRasterFile(rasterSrc)
	j := NewJsonFile(jsonSrc)
	outputs.Add("raster", &r)
	outputs.Add("summary", &j)

	if err := WriteOutputs(outputs, dir, nil); err != nil {
		t.Fatal(err)
	}

	if _, err := os.Stat(filepath.Join(dir, "outputs", "raster", "result.tif")); err != nil {
		t.Fatal("raster output not found")
	}
	if _, err := os.Stat(filepath.Join(dir, "outputs", "summary", "summary.json")); err != nil {
		t.Fatal("summary output not found")
	}
}

func TestWriteUnitOutput(t *testing.T) {
	dir := setupWorkDir(t)
	if err := WriteOutputs(unitOutput{}, dir, nil); err != nil {
		t.Fatal(err)
	}
	entries, _ := os.ReadDir(filepath.Join(dir, "outputs"))
	if len(entries) != 0 {
		t.Fatalf("expected empty outputs, got %d entries", len(entries))
	}
}

func TestFilenamePreservation(t *testing.T) {
	dir := setupWorkDir(t)
	src := filepath.Join(dir, "my_custom_name.geojson")
	os.WriteFile(src, []byte("geojson data"), 0644)

	result := NewVectorFile(src)
	if err := WriteOutputs(&result, dir, nil); err != nil {
		t.Fatal(err)
	}

	if _, err := os.Stat(filepath.Join(dir, "outputs", "vector", "my_custom_name.geojson")); err != nil {
		t.Fatal("expected original filename preserved")
	}
}

// ---------------------------------------------------------------------------
// ReadBlockManifest tests
// ---------------------------------------------------------------------------

func TestReadManifestNone(t *testing.T) {
	dir := setupWorkDir(t)
	t.Setenv("SPADE_BLOCK_MANIFEST", "")
	result := ReadBlockManifest(dir)
	if result != nil {
		t.Fatalf("expected nil, got %v", result)
	}
}

func TestReadManifestBlockYaml(t *testing.T) {
	dir := setupWorkDir(t)
	t.Setenv("SPADE_BLOCK_MANIFEST", "")
	content := "id: test.block\noutputs:\n  raster:\n    type: file\n    format: GeoTIFF\n"
	os.WriteFile(filepath.Join(dir, "block.yaml"), []byte(content), 0644)

	result := ReadBlockManifest(dir)
	if result == nil {
		t.Fatal("expected non-nil result")
	}
	if _, ok := result["raster"]; !ok {
		t.Fatal("expected 'raster' key")
	}
}

func TestReadManifestEnvVar(t *testing.T) {
	dir := setupWorkDir(t)
	manifestPath := filepath.Join(dir, "external_block.yaml")
	content := "outputs:\n  output:\n    type: file\n"
	os.WriteFile(manifestPath, []byte(content), 0644)

	t.Setenv("SPADE_BLOCK_MANIFEST", manifestPath)
	result := ReadBlockManifest(dir)
	if result == nil {
		t.Fatal("expected non-nil result")
	}
	if _, ok := result["output"]; !ok {
		t.Fatal("expected 'output' key")
	}
}

func TestReadManifestEnvVarTakesPrecedence(t *testing.T) {
	dir := setupWorkDir(t)

	// block.yaml in work dir
	cwdContent := "outputs:\n  cwd_output:\n    type: file\n"
	os.WriteFile(filepath.Join(dir, "block.yaml"), []byte(cwdContent), 0644)

	// external manifest
	extPath := filepath.Join(dir, "external.yaml")
	extContent := "outputs:\n  env_output:\n    type: file\n"
	os.WriteFile(extPath, []byte(extContent), 0644)

	t.Setenv("SPADE_BLOCK_MANIFEST", extPath)
	result := ReadBlockManifest(dir)
	if result == nil {
		t.Fatal("expected non-nil result")
	}
	if _, ok := result["env_output"]; !ok {
		t.Fatal("expected 'env_output' key")
	}
	if _, ok := result["cwd_output"]; ok {
		t.Fatal("should not contain 'cwd_output' — env var takes precedence")
	}
}
